### PR TITLE
add etag functionality for GET `/jobs/{id}/tasks`

### PR DESCRIPTION
**Pull request by justinpjose**
### What
**Describe what you have changed and why.**
- Ensure that we check the If-Match header in the request if it exists
- Ensure ETag is generated after retrieving data from database and checks this with the `IfMatch` header value if set and not set to `*`
- Ensure we set `ETag` header in the response back
- Add unit test and component test

### How to review
**Describe the steps required to test the changes.**
- Check if the test passes
- Check if the code changes make sense

### Who can review
**Describe who worked on the changes, so that other people can review.**
- Anyone

### CODE REVIEW

File: api/tasks_test.go [hunk 1]
```diff
@@ -502,6 +516,177 @@ func TestGetTasksHandler(t *testing.T) {
...
+		})
+	})
+
+	Convey("Given valid etag", t, func() {
+		httpClient := dpHTTP.NewClient()
+		apiInstance := api.Setup(mux.NewRouter(), dataStorerMock, &apiMock.AuthHandlerMock{}, taskNames, cfg, httpClient, &apiMock.IndexerMock{}, &apiMock.ReindexRequestedProducerMock{})
+
+		Convey("When request is made to get tasks", func() {
+			req := httptest.NewRequest(http.MethodGet, fmt.Sprintf("http://localhost:25700/jobs/%s/tasks?offset=%d&limit=%d", validJobID1, validOffset, validLimit), nil)
+			headers.SetIfMatch(req, `"024613bed9e9db8b730c94a98a420c2ef7e39a09"`)
```
Comment (cookel2) [inline api/tasks_test.go:534]: SetIfMatch returns an error so that needs to be handled.

File: api/tasks_test.go [hunk 2]
```diff
@@ -502,6 +516,177 @@ func TestGetTasksHandler(t *testing.T) {
...
+		})
+	})
+
+	Convey("Given empty etag", t, func() {
+		httpClient := dpHTTP.NewClient()
+		apiInstance := api.Setup(mux.NewRouter(), dataStorerMock, &apiMock.AuthHandlerMock{}, taskNames, cfg, httpClient, &apiMock.IndexerMock{}, &apiMock.ReindexRequestedProducerMock{})
+
+		Convey("When request is made to get tasks", func() {
+			req := httptest.NewRequest(http.MethodGet, fmt.Sprintf("http://localhost:25700/jobs/%s/tasks?offset=%d&limit=%d", validJobID1, validOffset, validLimit), nil)
+			headers.SetIfMatch(req, "")
```
Comment (cookel2) [inline api/tasks_test.go:582]: Please handle the SetIfMatch error here too.

File: api/tasks_test.go [hunk 3]
```diff
@@ -502,6 +516,177 @@ func TestGetTasksHandler(t *testing.T) {
...
+		})
+	})
+
+	Convey("Given If-Match header set to *", t, func() {
+		httpClient := dpHTTP.NewClient()
+		apiInstance := api.Setup(mux.NewRouter(), dataStorerMock, &apiMock.AuthHandlerMock{}, taskNames, cfg, httpClient, &apiMock.IndexerMock{}, &apiMock.ReindexRequestedProducerMock{})
+
+		Convey("When request is made to get tasks", func() {
+			req := httptest.NewRequest(http.MethodGet, fmt.Sprintf("http://localhost:25700/jobs/%s/tasks?offset=%d&limit=%d", validJobID1, validOffset, validLimit), nil)
+			headers.SetIfMatch(req, "*")
```
Comment (cookel2) [inline api/tasks_test.go:630]: Please handle the SetIfMatch error here too.

File: api/tasks_test.go [hunk 4]
```diff
@@ -502,6 +516,177 @@ func TestGetTasksHandler(t *testing.T) {
...
+		})
+	})
+
+	Convey("Given outdated or invalid etag", t, func() {
+		httpClient := dpHTTP.NewClient()
+		apiInstance := api.Setup(mux.NewRouter(), dataStorerMock, &apiMock.AuthHandlerMock{}, taskNames, cfg, httpClient, &apiMock.IndexerMock{}, &apiMock.ReindexRequestedProducerMock{})
+
+		Convey("When request is made to get tasks", func() {
+			req := httptest.NewRequest(http.MethodGet, fmt.Sprintf("http://localhost:25700/jobs/%s/tasks?offset=%d&limit=%d", validJobID1, validOffset, validLimit), nil)
+			headers.SetIfMatch(req, "invalid")
```
Comment (cookel2) [inline api/tasks_test.go:678]: And please handle the SetIfMatch error here too.

**Review by cookel2 — CHANGES_REQUESTED** [review]
Just a few comments.

File: api/tasks_test.go (hunk 1 above)
Comment (justinpjose) [reply api/tasks_test.go:534]: Done

File: api/tasks_test.go (hunk 2 above)
Comment (justinpjose) [reply api/tasks_test.go:582]: Done

File: api/tasks_test.go (hunk 3 above)
Comment (justinpjose) [reply api/tasks_test.go:630]: Done

**Review by cookel2 — APPROVED** [review]
looks good to me

**Pull request merged**